### PR TITLE
orientations: define entity_permutations for CG and DG elements

### PR DESCRIPTION
**Pull request by ksagiyam**
This PR is to attach `entity_permutations` `dict` to basic finite elements (such as `Lagrange`) that defines a DoF permutation (to be used in global vector assembly) for each possible entity orientation for each entity for each dimension. 
`entity_permutations` is set `None` if it is not yet implemented for that element, so the problem solving environment must handle this correctly if it ever tries to access this attribute.

This is intended to be the first step for FIAT to handle orientations.

### CODE REVIEW

File: FIAT/dual_set.py [hunk 1]
```diff
@@ -40,6 +41,11 @@ def get_entity_closure_ids(self):
     def get_entity_ids(self):
         return self.entity_ids
 
+    def get_entity_permutations(self):
```
Comment (dham) [inline FIAT/dual_set.py:44]: This needs serious documentation.

**Review by dham — CHANGES_REQUESTED** [review]

File: FIAT/dual_set.py (hunk 1 above)
Comment (ksagiyam) [reply FIAT/dual_set.py:44]: Doc string added!

**Comment by ksagiyam** [comment]
Added more detailed explanation on how permutations of `discontinuous_lagrange` element are computed.

File: FIAT/P0.py [hunk 2]
```diff
@@ -29,12 +30,23 @@ def __init__(self, ref_el):
...
+                perms = {o: [] for o in range(sym_size)}
             for entity in sorted(top[dim]):
                 entity_ids[dim][entity] = []
-
+                entity_permutations[dim][entity] = perms
         entity_ids[dim] = {0: [0]}
+        if isinstance(dim, tuple):
+            entity_permutations[dim][0] = {o: [0, ] for o in numpy.ndindex(sym_size)}
+        else:
+            entity_permutations[dim][0] = {o: [0, ] for o in range(sym_size)}
```
Comment (wence-) [inline FIAT/P0.py:47]: Style, singleton lists are spelt `[0]`.

File: FIAT/discontinuous_lagrange.py [hunk 3]
```diff
@@ -5,7 +5,110 @@
...
+    # dim = 2, degree = 3 (npoints = degree + 1)
+    #
+    #     facet ids    CG node numbers  DG node numbers
+    #    +
+    #    | \           3                1
+    #    |   \  0      2 6              8 3
+    #  2 |     \       1 5 8            7 9 4
+    #    |       \     0 4 7 9          0 5 6 2
+    #    +--------+
+    #        1
```
Comment (wence-) [inline FIAT/discontinuous_lagrange.py:31]: This is very confusing.

"DG degrees of freedom are numbered to geometrically coincide with CG degrees of freedom."

But note that the picture is wrong, because the reference cell is:

```
2
| \
|  \
1   0
|      \
|         \
0----2----1
```

File: FIAT/discontinuous_lagrange.py [hunk 4]
```diff
@@ -5,7 +5,110 @@
...
+    # a.shape[0] = number of DoFs
+    # a.shape[1] = dim + 1
+    # sum of each row = degree (bary centric lattice coordinates)
+    index_perms = make_index_permutations(dim + 1)
+    # index_perms = [[0, 1, 2],
+    #                [0, 2, 1],
+    #                [1, 0, 2],
+    #                [1, 2, 0],
+    #                [2, 0, 1],
+    #                [2, 1, 0]]
```
Comment (wence-) [inline FIAT/discontinuous_lagrange.py:52]: These bits just look like commented out code, can we have some exposition?

File: FIAT/discontinuous_lagrange.py [hunk 5]
```diff
@@ -5,7 +5,110 @@
...
+    g0 = dim - (a == 0).astype(int).sum(axis=1)
+    # g0 = [ 0, 1, 1, 0, 1, 2, 1, 1, 1, 0]  # 0 for vertices
+    #                                       # 1 for edges
+    #                                       # 2 for cell
```
Comment (wence-) [inline FIAT/discontinuous_lagrange.py:56]: "First separate by entity dimension." (or something)

File: FIAT/discontinuous_lagrange.py [hunk 6]
```diff
@@ -5,7 +5,110 @@
...
+    g1 = np.zeros_like(g0)
+    for d in range(dim + 1):
+        on_facet_d = (a[:, d] == 0).astype(int)
+        g1 += d * on_facet_d
```
Comment (wence-) [inline FIAT/discontinuous_lagrange.py:60]: Then separate by entity number.

File: FIAT/discontinuous_lagrange.py [hunk 7]
```diff
@@ -5,7 +5,110 @@
...
+    # Order first by g0, then by g1, and finally by a (lexicographically as in CG)
+    g0 = g0.reshape((a.shape[0], 1))
+    g1 = g1.reshape((a.shape[0], 1))
+    dg_cg_map = np.lexsort(np.transpose(np.concatenate((a, g1, g0), axis=1)))
+    # dg_cg_map = [ 0, 3, 9, 6, 8, 4, 7, 1, 2, 5]
+    cg_dg_map = np.empty_like(dg_cg_map)
+    for i, im in enumerate(dg_cg_map):
+        cg_dg_map[im] = i
+    # cg_dg_map = [ 0, 7, 8, 1, 5, 9, 3, 6, 4, 2]
+    perms = {}
```
Comment (wence-) [inline FIAT/discontinuous_lagrange.py:78]: This isn't computing a permutation between the DG element and an equivalent CG element, but rather the permutation to a lattice on the cell.

File: FIAT/discontinuous_lagrange.py [hunk 8]
```diff
@@ -5,7 +5,110 @@
...
+        # 8 7
+        # 6 5 4
+        # 3 2 1 0
+        #
+        # In this case we will have:
+        #
+        # perm                       = [3, 6, 8, 9, 2, 5, 7, 1, 4, 0]
+        # cg_dg_map[perm]            = [1, 3, 4, 2, 8, 9, 6, 7, 5, 0]
+        # cg_dg_map[perm][dg_cg_map] = [1, 2, 0, 6, 5, 8, 7, 3, 4, 9]
+        #
```
Comment (wence-) [inline FIAT/discontinuous_lagrange.py:108]: TODO: add comment noting that the sane thing to do is to just number DG dofs on a lattice.

File: FIAT/finite_element.py [hunk 9]
```diff
@@ -60,6 +60,11 @@ def entity_closure_dofs(self):
         freedom on the closure of those entities for the finite element."""
         return self.dual.get_entity_closure_ids()
 
+    def entity_permutations(self):
+        """Return the map of topological entities to the map of orientations
+        to permutation lists for the finite element."""
```
Comment (wence-) [inline FIAT/finite_element.py:65]: Can you just paste the dualset docstring here?

```
entity_permutations.__doc__ = DualSet.get_entity_permutations.__doc__
```

File: FIAT/finite_element.py [hunk 10]
```diff
@@ -242,3 +247,67 @@ def entity_support_dofs(elem, entity_dim):
...
+    asub = a[:, j]
+    i0 = i
+    if j == 0:
+        asub[i0: i0 + 1] = m
+        return
+    for v in range(m + 1):
+        i1 = i0 + _nhk(j, m - v)
+        asub[i0: i1] = v
+        _set_barycentric_lattice_coordinates(a, i0, j - 1, m - v)
+        i0 = i1
```
Comment (wence-) [inline FIAT/finite_element.py:269]: A bit hard to follow...

File: FIAT/finite_element.py [hunk 11]
```diff
@@ -242,3 +247,67 @@ def entity_support_dofs(elem, entity_dim):
...
+    n = numpy.math.factorial(len(lst) - 1)
+    for v in lst:
+        i1 = i0 + n
+        asub[i0: i1] = v
+        _set_index_permutations(a, i0, j + 1, [item for item in lst if item != v])
+        i0 = i1
+
+
+def make_index_permutations(n):
+    """Make all ordered permutations of [0, 1, ..., n - 1]"""
```
Comment (wence-) [inline FIAT/finite_element.py:310]: Is this equivalent to:

```
sorted(itertools.permutations(range(n))
```
?

File: FIAT/finite_element.py [hunk 12]
```diff
@@ -242,3 +247,67 @@ def entity_support_dofs(elem, entity_dim):
...
+     [1, 0, 2],
+     [1, 1, 1],
+     [1, 2, 0],
+     [2, 0, 1],
+     [2, 1, 0],
+     [3, 0, 0]]
+    """
+    a = numpy.zeros((_nhk(dim + 1, n), dim + 1), dtype=int)
+    _set_barycentric_lattice_coordinates(a, 0, dim, n)
+    return a
```
Comment (wence-) [inline FIAT/finite_element.py:293]:
```
sorted(mis(dim+1, n))
```

?

File: FIAT/lagrange.py [hunk 13]
```diff
@@ -6,6 +6,65 @@
...
+    triangular cell:
+
+       +                    +
+       | \                  | \
+       2   0               47   42
+       |     \              |     \
+       +--1---+             +--43--+
+    FIAT canonical     Mapped example physical cell
+
+    Suppose that the cone of the physical cell is given by:
```
Comment (wence-) [inline FIAT/lagrange.py:28]: cone isn't a word in the FIAT vocabulary.

File: FIAT/P0.py (hunk 2 above)
Comment (ksagiyam) [reply FIAT/P0.py]: Thanks. Fixed.

File: FIAT/discontinuous_lagrange.py (hunk 3 above)
Comment (ksagiyam) [reply FIAT/discontinuous_lagrange.py]: Fixed.

File: FIAT/discontinuous_lagrange.py (hunk 4 above)
Comment (ksagiyam) [reply FIAT/discontinuous_lagrange.py]: I think it's now easier to read.

File: FIAT/discontinuous_lagrange.py (hunk 5 above)
Comment (ksagiyam) [reply FIAT/discontinuous_lagrange.py]: Done.

File: FIAT/discontinuous_lagrange.py (hunk 6 above)
Comment (ksagiyam) [reply FIAT/discontinuous_lagrange.py:79]: Done.

File: FIAT/discontinuous_lagrange.py (hunk 7 above)
Comment (ksagiyam) [reply FIAT/discontinuous_lagrange.py]: Renamed: `dg_to_lattice` and `lattice_to_dg`.

File: FIAT/discontinuous_lagrange.py (hunk 8 above)
Comment (ksagiyam) [reply FIAT/discontinuous_lagrange.py:134]: Done.

File: FIAT/finite_element.py (hunk 9 above)
Comment (ksagiyam) [reply FIAT/finite_element.py]: Done.

File: FIAT/finite_element.py (hunk 10 above)
Comment (ksagiyam) [reply FIAT/finite_element.py]: These weird bits have all been removed. Thanks!

File: FIAT/finite_element.py (hunk 11 above)
Comment (ksagiyam) [reply FIAT/finite_element.py]: Yes!

File: FIAT/finite_element.py (hunk 12 above)
Comment (ksagiyam) [reply FIAT/finite_element.py]: Yes! Actually I had to do `sorted(mis(dim+1, n))[:, ::-1]` to flip along axis 1, but that was just to use this array with `lexsort`.

File: FIAT/lagrange.py (hunk 13 above)
Comment (ksagiyam) [reply FIAT/lagrange.py]: Fixed. Thanks.

**Review by dham — APPROVED** [review]
